cmd/snap-bootstrap/cmd_scan_disk.go

```diff
@@ -0,0 +1,376 @@
+ *
+ * LABEL="ubuntu_core_partitions_end"
+ * ```
+ *
+ * See
+ * core-initrd/latest/factory/usr/lib/udev/rules.d/90-ubuntu-core-partitions.rules
+ * for implementation.
+ *
+ * Note that symlink /dev/disk/ubuntu/disk can be expected by
+ * snap-bootstrap. In that case, snap-initramfs-mounts.service should
+ * have:
+ *
+ * ```
+ * BindsTo=dev-disk-ubuntu-disk.device
+ * After=dev-disk-ubuntu-disk.device
+ * ```
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/snapcore/snapd/boot"
+	"github.com/snapcore/snapd/bootloader/efi"
+	"github.com/snapcore/snapd/cmd/snap-bootstrap/blkid"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil/kcmdline"
+)
+
+var (
+	efiReadVarString = efi.ReadVarString
+)
+
+func init() {
+	const (
+		short = "Verify that a disk is the booting disk"
+		long  = "This tool is expected to be called from udev"
+	)
+
+	addCommandBuilder(func(parser *flags.Parser) {
+		if _, err := parser.AddCommand("scan-disk", short, long, &cmdScanDisk{}); err != nil {
+			panic(err)
+		}
+	})
+}
+
+type cmdScanDisk struct{}
+
+func (c *cmdScanDisk) Execute([]string) error {
+	return ScanDisk(os.Stdout)
+}
+
+type Partition struct {
+	Name string
+	UUID string
+}
+
+func isGpt(probe blkid.AbstractBlkidProbe) bool {
+	pttype, err := probe.LookupValue("PTTYPE")
+	if err != nil {
+		return false
+	}
+	return pttype == "gpt"
+}
+
+func probePartitions(node string) ([]Partition, error) {
+	probe, err := blkid.NewProbeFromFilename(node)
+	if err != nil {
+		return nil, err
+	}
+	defer probe.Close()
+
+	probe.EnablePartitions(true)
+	probe.SetPartitionsFlags(blkid.BLKID_PARTS_ENTRY_DETAILS)
+	probe.EnableSuperblocks(true)
+
+	if err := probe.DoSafeprobe(); err != nil {
+		return nil, err
+	}
+
+	if !isGpt(probe) {
+		return nil, nil
+	}
+
+	partitions, err := probe.GetPartitions()
+	if partitions == nil {
+		return nil, err
+	}
+
+	ret := make([]Partition, 0)
+	for _, partition := range partitions.GetPartitions() {
+		label := partition.GetName()
+		uuid := partition.GetUUID()
+		fmt.Fprintf(os.Stderr, "Found partition %s %s\n", label, uuid)
+		ret = append(ret, Partition{label, uuid})
+	}
+
+	return ret, nil
+}
+
+func samePath(a, b string) (bool, error) {
+	aSt, err := os.Stat(a)
+	if err != nil {
+		return false, err
+	}
+	bSt, err := os.Stat(b)
+	if err != nil {
+		return false, err
+	}
+	return os.SameFile(aSt, bSt), nil
+}
+
+func scanDiskNode(output io.Writer, node string) error {
+	/*
+	 * We need to find out if the give contains the ESP that was booted.
+	 * The boot loader will set LoaderDevicePartUUID. We will need to scan all
+	 * the partitions for that UUID.
+	 */
+	fmt.Fprintf(os.Stderr, "Scanning disk %s\n", node)
+	fallback := false
+	var fallbackPartition string
+	bootUUID, _, err := efiReadVarString("LoaderDevicePartUUID-4a67b082-0a4c-41cf-b6c7-440b29bb8c4f")
```

Comment:
we have boot.FindPartitionUUIDForBootedKernelDisk for reading this as well I think

Reply:
@valentindavid did you see this comment/question?